Resteasy Reactive: Support produces Multipart FormData

Sgitario opened the pull request:
Add RESTEasy Reactive support to produce Multipart Form data.
- Updated RESTEasy Reactive documentation
- Supported both `@RestForm` and `@FormParam` annotations
- Added test coverage for text/plan, json and xml serialization
- Implementation is based on Resteasy Classic documentation

Resolves https://github.com/quarkusio/quarkus/issues/21009

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java:
```diff
@@ -0,0 +1,213 @@
+    public void writeTo(Object o, Class<?> aClass, Type type, Annotation[] annotations, MediaType mediaType,
+            MultivaluedMap<String, Object> httpHeaders, OutputStream outputStream)
+            throws IOException, WebApplicationException {
+        if (o instanceof String) {
```
Does this writer still get used when the return type is string?

Sgitario replied:
No, I deleted this as the writer is not always configured, but only when the multipart form data is used.

geoand replied:
But my question still remains: When the writer is configured, is it being used to write out strings? If so, that is wrong.

Sgitario replied:
No, this is the final version of the writer: https://github.com/quarkusio/quarkus/pull/21038/files#diff-e8ce68a54c45a2a5550a9849588de76e36a5e0ba39530a2adab777d97978a20dR32

geoand replied:
And you have verified that returning a `String` continues to work properly even when this multipart writer has been configured do to it's use by some other method?

Sgitario replied:
Yes, I added this test case: https://github.com/quarkusio/quarkus/pull/21038/files#diff-136f284323fc8e3022c3f73dce4ed1d4e1ce5d4adf1c6e9affe221f6b0c6befdR48

geoand replied:
Great, thanks!

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -474,6 +472,49 @@ Moreoever if `quarkus.http.body.delete-uploaded-files-on-end` is set to true, Qu
+}
+----
+
+WARNING: Returning Multipart data has been limited to be only working for blocking endpoints in order to avoid the event loop thread to deal with files. 
```
```suggestion
WARNING: For the time being, returning Multipart data is limited to be blocking endpoints. 
```

Sgitario replied:
Updated

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/FormDataOutputMapperGenerator.java:
```diff
@@ -0,0 +1,251 @@
+     *
+     * The generated mapper would look like:
+     * public class FormData_generated_mapper implements MultipartOutputInjectionTarget {
+     * public FormDataOutput mapFrom(Object var1) {
+     * FormDataOutput var2 = new FormDataOutput();
+     * FormData var4 = (FormData)var1;
+     * File var3 = var4.data;
+     * MultipartSupport.addPartItemToFormDataOutput(var2, "file", "application/octet-stream", var3);
+     * File var5 = var4.text;
+     * MultipartSupport.addPartItemToFormDataOutput(var2, "text", "text/plain", var5);
+     * return var2;
```
This could use better formatting as it's not readable

Sgitario replied:
Ups, this is wrong. Thanks for spotting. Fixed.

geoand commented on independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java:
```diff
@@ -535,6 +542,12 @@ private ResourceMethod createResourceMethod(ClassInfo currentClassInfo, ClassInf
 
+            if (returnsMultipart && !blocking) {
+                throw new DeploymentException(
+                        "Endpoints with '@Produces(MediaType.MULTIPART_FORM_DATA)' can only be used on non blocking methods. Offending method is '"
```
```suggestion
                        "Endpoints that produce a Multipart result can only be used on non blocking methods. Offending method is '"
```

Sgitario replied:
Updated

geoand commented:
Let's only include this after https://github.com/quarkusio/quarkus/pull/21138 (which is very important) has gone in

geoand commented:
There is a conflict now that https://github.com/quarkusio/quarkus/pull/21138 went in

geoand reviewed (approved):
Thanks!

geoand commented:
Something keeps breaking in CI, so I think it's an actual problem that needs to be fixed.

Sgitario commented:
> Something keeps breaking in CI, so I think it's an actual problem that needs to be fixed.

The new writer was not registered for reflection and hence was failing when running on Native. Fixed.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java:
```diff
@@ -612,6 +613,9 @@ public void serverSerializers(ResteasyReactiveRecorder recorder,
         serverSerializersProducer.produce(new ServerSerialisersBuildItem(serialisers));
+
+        // conditional serialisers
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, MultipartMessageBodyWriter.class.getName()));
```
Why are we always registering this for reflection even if it's not going to be used?

Sgitario replied:
I tried to register conditionally in here: QuarkusServerEndpointIndexer.handleMultipartForReturnType, but didn't work. Let me try again.

Sgitario replied:
It should be fixed now.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java:
```diff
@@ -0,0 +1,169 @@
+    private static final String LINE_SEPARATOR = "\r\n";
+    private static final String BOUNDARY_PARAM = "boundary";
+
+    private Charset defaultCharset;
```
The way you are setting this, this needs to be `volatile`. 

But a better option would be to set it in the constructor, no?

Sgitario replied:
Changed

geoand reviewed (changes requested):
See my latest comment

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java:
```diff
@@ -0,0 +1,165 @@
+    private static final String LINE_SEPARATOR = "\r\n";
+    private static final String BOUNDARY_PARAM = "boundary";
+
+    private volatile Charset defaultCharset;
```
I guess I was not clear enough.

`volatile` is only needed if the field is **not** set in the constructor, but in a method that can be called by other threads.

If you set the field in the constructor, `volatile` is not needed and the field can be `final`.

Sgitario replied:
Ok, it makes more sense now. Changed.

Sgitario committed:
Resteasy Reactive: Support produces Multipart FormData

Add RESTEasy Reactive support to produce Multipart Form data.
- Updated RESTEasy Reactive documentation
- Supported both `@RestForm` and `@FormParam` annotations
- Added test coverage for text/plan, json and xml serialization
- Implementation is based on Resteasy Classic documentation
